libdrgn/arch_s390.c

```diff
@@ -0,0 +1,327 @@
+		err = DRGN_UNKNOWN_RELOCATION_TYPE(r_type);
+		break;
+	}
+	return err;
+}
+
+static struct drgn_error *
+unwind_get_pt_regs_sizes(struct drgn_program *prog,
+			 uint64_t *gprs_offset, uint64_t *psw_offset,
+			 uint64_t *addr_offset, uint64_t *mask_offset,
+			 uint64_t *frame_size, uint64_t *ptregs_size)
+{
+	struct drgn_qualified_type pt_type, frame_type, psw_type;
+	struct drgn_type_member *member;
+	struct drgn_error *err;
+
+	err = drgn_program_find_type(prog, "struct stack_frame", NULL, &frame_type);
+	if (err)
+		return err;
+	*frame_size = drgn_type_size(frame_type.type);
+
+	err = drgn_program_find_type(prog, "struct pt_regs", NULL, &pt_type);
+	if (err)
+		return err;
+	*ptregs_size = drgn_type_size(pt_type.type);
+
+	err = drgn_type_find_member(pt_type.type, "gprs", &member, gprs_offset);
+	if (err)
+		return err;
+
+	err = drgn_type_find_member(pt_type.type, "psw", &member, psw_offset);
+	if (err)
+		return err;
+
+	err = drgn_program_find_type(prog, "psw_t", NULL, &psw_type);
+	if (err)
+		return err;
+
+	err = drgn_type_find_member(psw_type.type, "addr", &member, addr_offset);
+	if (err)
+		return err;
+
+	return drgn_type_find_member(psw_type.type, "mask", &member, mask_offset);
+}
+
+/*
+ * s390 puts a stackframe + stack pt_regs on the kernel stack when entering
+ * program check handler, irq handlers or the system call handler.
+ */
+static struct drgn_error *
+unwind_check_pt_regs(struct drgn_program *prog,
+		     struct drgn_register_state *regs,
+		     struct drgn_register_state **ret)
+{
+	struct optional_uint64 tmp = drgn_register_state_get_u64(prog, regs, r15);
+	uint64_t psw_offset, mask_offset, addr_offset, gprs_offset, bc_offset;
+	uint64_t nextbc, frame_size, ptregs_size, *pswa, *pswm;
+	struct drgn_qualified_type frame_type;
+	struct drgn_type_member *member;
+	struct drgn_error *err;
+
+	if (!tmp.has_value || !tmp.value)
+		return &drgn_stop;
+
+	err = unwind_get_pt_regs_sizes(prog, &gprs_offset, &psw_offset, &addr_offset,
+				       &mask_offset, &frame_size, &ptregs_size);
+	if (err)
+		return err;
+
+	uint8_t *frame = malloc(frame_size + ptregs_size);
+	if (!frame)
+		return &drgn_enomem;
+
+	err = drgn_program_read_memory(prog, frame, tmp.value,
+				       frame_size + ptregs_size, false);
+	if (err)
+		goto out;
+
+	struct drgn_register_state *regs2 = drgn_register_state_create(bc, false);
+	if (!regs2)
+		goto out;
+
+	uint8_t *ptregs = frame + frame_size;
+	pswa = (uint64_t *)(ptregs + (psw_offset + addr_offset) / 8);
+	pswm = (uint64_t *)(ptregs + (psw_offset + mask_offset) / 8);
+
+        uint64_t *gprs = (uint64_t *)(ptregs + gprs_offset / 8);
+	drgn_register_state_set_range_from_buffer(regs2, r0, r15, gprs);
+	drgn_register_state_set_from_buffer(regs2, pswa, pswa);
+	drgn_register_state_set_from_buffer(regs2, pswm, pswm);
+	drgn_register_state_set_pc_from_register(prog, regs2, pswa);
+	drgn_register_state_set_from_buffer(regs2, bc, &gprs[15]);
+
+	*ret = regs2;
+
+	if (*pswm & (1UL << 48)) {
```

Comment:
I actually prefer that we include the userspace frame; it can be useful to know what the original syscall arguments were. The fallback unwind callback should check whether it gets a fault trying to read memory and stop unwinding.